provider/oci/images_test.go

```diff
@@ -0,0 +1,393 @@
+		i.testCompartment, []map[string]string{
+			map[string]string{
+				"id":      "fakeUbuntu1",
+				"os":      "Canonical Ubuntu",
+				"name":    "Canonical-Ubuntu-14.04-2018.01.11-0",
+				"version": "14.04",
+			},
+			map[string]string{
+				"id":      fakeUbuntuIDSecond,
+				"os":      "Canonical Ubuntu",
+				"name":    "Canonical-Ubuntu-14.04-2018.01.12-0",
+				"version": "14.04",
+			},
+			map[string]string{
+				"id":      "fakeCentOS",
+				"os":      "CentOS",
+				"name":    "CentOS-7-2017.10.19-0",
+				"version": "7",
+			},
+		})
+
+	shapesRequestUbuntu, shapesResponseUbuntu := makeShapesRequestResponse(
+		i.testCompartment, "fakeUbuntu1", []string{"VM.Standard2.1", "VM.Standard1.2"})
+
+	shapesRequestUbuntuSecond := ociCore.ListShapesRequest{
+		CompartmentId: &i.testCompartment,
+		ImageId:       &fakeUbuntuIDSecond,
+	}
+
+	shapesRequestCentOS, shapesResponseCentOS := makeShapesRequestResponse(
+		i.testCompartment, "fakeCentOS", []string{"VM.Standard1.2"})
+
+	gomock.InOrder(
+		compute.EXPECT().ListImages(context.Background(), listImageRequest).Return(listImageResponse, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestUbuntu).Return(shapesResponseUbuntu, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestUbuntuSecond).Return(shapesResponseUbuntu, nil),
+		compute.EXPECT().ListShapes(context.Background(), shapesRequestCentOS).Return(shapesResponseCentOS, nil),
+	)
+
+	imgCache, err := oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+	c.Assert(imgCache, gc.NotNil)
+	c.Check(len(imgCache.ImageMap()), gc.Equals, 2)
+
+	imageMap := imgCache.ImageMap()
+	c.Check(len(imageMap["trusty"]), gc.Equals, 2)
+	c.Check(len(imageMap["centos7"]), gc.Equals, 1)
+
+	timeStamp, _ := time.Parse("2006.01.02", "2018.01.12")
+
+	// Check that the first image in the array is the newest one
+	c.Assert(imageMap["trusty"][0].Version.TimeStamp, gc.Equals, timeStamp)
+
+	// Check that InstanceTypes are set
+	c.Assert(len(imageMap["trusty"][0].InstanceTypes), gc.Equals, 2)
+	c.Assert(len(imageMap["centos7"][0].InstanceTypes), gc.Equals, 1)
+}
+
+func (i *imagesSuite) TestRefreshImageCacheStaleCache(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	compute := ocitesting.NewMockOCIComputeClient(ctrl)
+	defer ctrl.Finish()
+
+	compute.EXPECT().ListImages(gomock.Any(), gomock.Any()).Return(ociCore.ListImagesResponse{}, nil).Times(2)
+
+	imgCache, err := oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+	c.Assert(imgCache, gc.NotNil)
+
+	fromCache, err := oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
+	c.Check(imgCache, gc.DeepEquals, fromCache)
+
+	now := time.Now()
+
+	// No need to check the value. gomock will assert if ListImages
+	// is not called twice
+	fromCache.SetLastRefresh(now.Add(-31 * time.Minute))
+	_, err = oci.RefreshImageCache(compute, &i.testCompartment)
+	c.Assert(err, gc.IsNil)
```

Comment:
Can you actually break this test into 2 tests?
One test that calling twice without changing SetLastRefresh only has a single call.
And then one test that after setting the LastRefresh you see it called 2x?

Otherwise, this test would pass if you accidentally inverted the logic around checking last refresh. (if calling SetLastRefresh caused us *not* to refresh the image cache, this test still passes because the first 2 calls both trigger a call, and then the last one doesn't.)

Reply:
Done